### PR TITLE
port protect villages from being overbuilt

### PR DESCRIPTION
**Pull request by Raycoms**
Port

Changes proposed in this pull request:

Remove entities after constructing
Protect villages

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 1]
```diff
@@ -422,11 +446,17 @@ public boolean isCorrectMD5(final String otherMD5)
     /**
      * Reads a template from an inputstream.
      */
-    private static Template readTemplateFromStream(final InputStream stream) throws IOException
+    private static Template readTemplateFromStream(final InputStream stream, final DataFixer fixer) throws IOException
     {
         final NBTTagCompound nbttagcompound = CompressedStreamTools.readCompressed(stream);
+
+        if (!nbttagcompound.hasKey("DataVersion", 99))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:453]: MAJOR Assign this magic number 99 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 2]
```diff
@@ -422,11 +446,17 @@ public boolean isCorrectMD5(final String otherMD5)
...
+        {
+            nbttagcompound.setInteger("DataVersion", 500);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:455]: MAJOR Assign this magic number 500 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 51 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 36 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L127](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/structures/helpers/Structure.java#L127): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L236](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/structures/helpers/Structure.java#L236): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L249](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/structures/helpers/Structure.java#L249): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L59](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L59): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L793](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L793): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L52](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L52): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L177](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L177): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L190](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L190): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L71](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L71): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/c9516ea2f2db4a40e8f632f1fb80d73688ee750e/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 51 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 36 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L127](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/structures/helpers/Structure.java#L127): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L236](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/structures/helpers/Structure.java#L236): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L249](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/structures/helpers/Structure.java#L249): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L59](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L59): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L793](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L793): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L53](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L178](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L178): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L191](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L191): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L71](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L71): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/9f4801a7b15e22100a8a48ce7b0211c888878838/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Asherslab** [comment]
:shipit: I'm really starting to hate those Sonar blockers

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/1418/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**